examples/src/main/java/com/datastax/oss/driver/examples/concurrent/LimitConcurrencyCustom.java

```diff
@@ -0,0 +1,127 @@
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.CqlSessionBuilder;
+import com.datastax.oss.driver.api.core.cql.AsyncResultSet;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Creates a keyspace and tables, and loads data using Async API into them.
+ *
+ * <p>This example makes usage of a {@link CqlSession#executeAsync(String)} method, which is
+ * responsible for executing requests in a non-blocking way. It uses {@link ExecutorService} to
+ * limit number of concurrent request to CONCURRENCY_LEVEL.
+ *
+ * <p>Preconditions: - a Cassandra session is running and accessible through the contacts points
+ * identified by basic.contact-points (see application.conf)
+ *
+ * <p>Side effects: - creates a new keyspace "examples" in the session. If a keyspace with this name
+ * already exists, it will be reused; - creates a table "examples.tbl_sample_kv". If it exist
+ * already, it will be reused; - inserts a TOTAL_NUMBER_OF_INSERTS of rows into the table.
+ *
+ * @see <a href="http://datastax.github.io/java-driver/manual/">Java driver online manual</a>
+ */
+public class LimitConcurrencyCustom {
+  private static final int CONCURRENCY_LEVEL = 32;
+  private static final int TOTAL_NUMBER_OF_INSERTS = 10_000;
+
+  public static void main(String[] args) throws InterruptedException {
+
+    try (CqlSession session = new CqlSessionBuilder().build()) {
+      createSchema(session);
+      insertConcurrent(session);
+    }
+  }
+
+  private static void insertConcurrent(CqlSession session) throws InterruptedException {
+    PreparedStatement pst =
+        session.prepare(
+            insertInto("examples", "tbl_sample_kv")
+                .value("id", bindMarker("id"))
+                .value("value", bindMarker("value"))
+                .build());
+    // Create CountDownLatch that wait for completion of all pending requests
+    CountDownLatch requestLatch = new CountDownLatch(TOTAL_NUMBER_OF_INSERTS);
+    // Executor service with CONCURRENCY_LEVEL number of threads that states an upper limit
+    // on number of request in progress.
+    ExecutorService executor = Executors.newFixedThreadPool(CONCURRENCY_LEVEL);
+
+    // Used to track number of total inserts
+    AtomicInteger insertsCounter = new AtomicInteger();
+    // Used to track threads that were involved in a processing
+    Set<String> threads = new ConcurrentSkipListSet<>();
+
+    // For every i we will insert a record to db
+    for (int i = 0; i < TOTAL_NUMBER_OF_INSERTS; i++) {
```

Comment:
This is submitting 10,000 requests at once. I think what you need instead is a `Semaphore` with, say, 500 permits, and before submitting a request, you need to acquire 1 permit and once the request is executed, you release 1 permit. That way you would maintain a maximum threshold of 500 in-flight requests at all times.

Reply:
Good Idea, I've added a semaphore to limit the number of in-flight.